File: src/main/kotlin/com/vitiquest/peerreview/jira/JiraClient.kt
package com.vitiquest.peerreview.jira

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.vitiquest.peerreview.bitbucket.BitbucketClient
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.Base64

class JiraClient(
    baseUrl: String,
    private val email: String,
    private val apiToken: String
) {

    private val mapper = jacksonObjectMapper()
    private val http = BitbucketClient.http
    private val normalizedBaseUrl = baseUrl.trim().trimEnd('/')

    fun getIssue(issueKey: String): JiraIssue {
        val encodedKey = URLEncoder.encode(issueKey, StandardCharsets.UTF_8)
        val response: JiraIssueResponse = mapper.readValue(
            get("$normalizedBaseUrl/rest/api/3/issue/$encodedKey?fields=summary")
        )
        return JiraIssue(key = response.key, summary = response.fields.summary)
    }

    fun addPlainTextComment(issueKey: String, text: String) {
        val encodedKey = URLEncoder.encode(issueKey, StandardCharsets.UTF_8)
        val payload = mapper.writeValueAsString(buildCommentPayload(text))
        post("$normalizedBaseUrl/rest/api/3/issue/$encodedKey/comment", payload)
    }

    fun searchUsers(query: String): List<JiraUser> {
        val encodedQuery = URLEncoder.encode(query, StandardCharsets.UTF_8)
        return mapper.readValue(get("$normalizedBaseUrl/rest/api/3/user/search?query=$encodedQuery"))
    }

    fun assignIssue(issueKey: String, accountId: String) {
        val encodedKey = URLEncoder.encode(issueKey, StandardCharsets.UTF_8)
        val payload = mapper.writeValueAsString(mapOf("accountId" to accountId))
        put("$normalizedBaseUrl/rest/api/3/issue/$encodedKey/assignee", payload)
    }

    private fun get(url: String): String {
        val request = Request.Builder()
            .url(url)
            .header("Authorization", authHeader())
            .header("Accept", "application/json")
            .get()
            .build()
        return execute(request)
    }

    private fun post(url: String, jsonBody: String): String {
        val body = jsonBody.toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(url)
            .header("Authorization", authHeader())
            .header("Accept", "application/json")
            .post(body)
            .build()
        return execute(request)
    }

    private fun put(url: String, jsonBody: String): String {
        val body = jsonBody.toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(url)
            .header("Authorization", authHeader())
            .header("Accept", "application/json")
            .put(body)
            .build()
        return execute(request)
    }

    private fun execute(request: Request): String {
        http.newCall(request).execute().use { response ->
            val body = response.body?.string() ?: ""
            if (!response.isSuccessful) {
                val hint = when (response.code) {
                    401 -> " — Check the Atlassian email/API token in Settings → PR Pilot → JIRA."
                    403 -> " — The Atlassian account does not have permission for this JIRA project."
                    404 -> " — The issue does not exist or the Atlassian account lacks Browse permission " +
                           "for this project. Verify the issue key, and ensure the account in " +
                           "Settings → PR Pilot → JIRA is a member of the JIRA project."
                    else -> ""
                }
                throw IOException("JIRA API ${response.code}$hint [${request.method} ${request.url}]: ${body.take(400)}")
            }
            return body
        }
    }

    private fun authHeader(): String {
        val raw = "$email:$apiToken"
        val encoded = Base64.getEncoder().encodeToString(raw.toByteArray(StandardCharsets.UTF_8))
        return "Basic $encoded"
    }

    private fun buildCommentPayload(text: String): Map<String, Any> {
        val paragraphs = text.trim()
            .ifBlank { "Code Review Passed." }
            .split(Regex("\\R{2,}"))
            .map { paragraph ->
                val lines = paragraph.lines()
                val content = mutableListOf<Map<String, Any>>()
                lines.forEachIndexed { index, line ->
                    if (line.isNotEmpty()) {
                        content.add(mapOf("type" to "text", "text" to line))
                    }
                    if (index < lines.lastIndex) {
                        content.add(mapOf("type" to "hardBreak"))
                    }
                }
                mapOf(
                    "type" to "paragraph",
                    "content" to content.ifEmpty { listOf(mapOf("type" to "text", "text" to " ")) }
                )
            }

        return mapOf(
            "body" to mapOf(
                "type" to "doc",
                "version" to 1,
                "content" to paragraphs
            )
        )
    }
}

data class JiraIssue(
    val key: String,
    val summary: String
)

@JsonIgnoreProperties(ignoreUnknown = true)
private data class JiraIssueResponse(
    val key: String = "",
    val fields: JiraIssueFields = JiraIssueFields()
)

@JsonIgnoreProperties(ignoreUnknown = true)
private data class JiraIssueFields(
    val summary: String = ""
)

@JsonIgnoreProperties(ignoreUnknown = true)
data class JiraUser(
    val accountId: String = "",
    val displayName: String = "",
    val emailAddress: String? = null,
    val active: Boolean = true
)

File: vscode-plugin/src/jira/JiraClient.ts
import { httpGet, httpPost, httpPut } from '../utils/httpClient';
import { PullRequest } from '../models/PullRequest';

export interface JiraIssue {
  key: string;
  summary: string;
}

export interface JiraUser {
  accountId: string;
  displayName: string;
  emailAddress?: string;
  active: boolean;
}

interface JiraIssueResponse {
  key: string;
  fields: { summary: string };
}

const TIMEOUT = 30_000;

/**
 * Atlassian JIRA Cloud REST API client.
 * Uses Basic auth (email:apiToken, Base64-encoded).
 */
export class JiraClient {
  private readonly baseUrl: string;

  constructor(
    baseUrl: string,
    private readonly email: string,
    private readonly apiToken: string
  ) {
    this.baseUrl = baseUrl.trim().replace(/\/$/, '');
  }

  private authHeaders(): Record<string, string> {
    const credentials = Buffer.from(`${this.email}:${this.apiToken}`, 'utf8').toString('base64');
    return {
      'Authorization': `Basic ${credentials}`,
      'Accept': 'application/json',
      'Content-Type': 'application/json',
    };
  }

  private handleError(err: unknown, context: string): never {
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes('HTTP 401')) {
      throw new Error(`${msg}\nCheck the Atlassian email/API token in Settings → PR Pilot → JIRA.`);
    }
    if (msg.includes('HTTP 403')) {
      throw new Error(`${msg}\nThe Atlassian account does not have permission for this JIRA project.`);
    }
    if (msg.includes('HTTP 404')) {
      throw new Error(
        `${msg}\nThe issue does not exist or the account lacks Browse permission ` +
        `for this project. Verify the issue key and ensure the account is a member of the JIRA project.`
      );
    }
    throw err instanceof Error ? err : new Error(`${context}: ${msg}`);
  }

  async getIssue(issueKey: string): Promise<JiraIssue> {
    const url = `${this.baseUrl}/rest/api/3/issue/${encodeURIComponent(issueKey)}?fields=summary`;
    try {
      const body = await httpGet(url, this.authHeaders(), TIMEOUT);
      const parsed = JSON.parse(body) as Record<string, unknown>;
      // Some Jira instances return HTTP 200 with an errorMessages array instead of a proper 4xx.
      if (Array.isArray(parsed['errorMessages']) && (parsed['errorMessages'] as unknown[]).length > 0) {
        throw new Error(`HTTP 404: ${body}`);
      }
      const raw = parsed as unknown as JiraIssueResponse;
      return { key: raw.key, summary: raw.fields.summary };
    } catch (err) {
      this.handleError(err, `getIssue(${issueKey})`);
    }
  }

  async addPlainTextComment(issueKey: string, text: string): Promise<void> {
    const url = `${this.baseUrl}/rest/api/3/issue/${encodeURIComponent(issueKey)}/comment`;
    const payload = JSON.stringify(this.buildCommentPayload(text));
    try {
      await httpPost(url, payload, this.authHeaders(), TIMEOUT);
    } catch (err) {
      this.handleError(err, `addComment(${issueKey})`);
    }
  }

  async searchUsers(query: string): Promise<JiraUser[]> {
    const url = `${this.baseUrl}/rest/api/3/user/search?query=${encodeURIComponent(query)}`;
    try {
      return JSON.parse(await httpGet(url, this.authHeaders(), TIMEOUT)) as JiraUser[];
    } catch (err) {
      this.handleError(err, `searchUsers(${query})`);
    }
  }

  async assignIssue(issueKey: string, accountId: string): Promise<void> {
    const url = `${this.baseUrl}/rest/api/3/issue/${encodeURIComponent(issueKey)}/assignee`;
    try {
      await httpPut(url, JSON.stringify({ accountId }), this.authHeaders(), TIMEOUT);
    } catch (err) {
      this.handleError(err, `assignIssue(${issueKey})`);
    }
  }

  private buildCommentPayload(text: string): Record<string, unknown> {
    const paragraphs = (text.trim() || 'Code Review Passed.')
      .split(/\r?\n\r?\n/)
      .map((paragraph) => {
        const lines = paragraph.split(/\r?\n/);
        const content: Array<Record<string, unknown>> = [];
        lines.forEach((line, index) => {
          if (line.length > 0) {
            content.push({ type: 'text', text: line });
          }
          if (index < lines.length - 1) {
            content.push({ type: 'hardBreak' });
          }
        });
        return {
          type: 'paragraph',
          content: content.length > 0 ? content : [{ type: 'text', text: ' ' }],
        };
      });

    return {
      body: {
        type: 'doc',
        version: 1,
        content: paragraphs,
      },
    };
  }
}
